tests/functional/test_credentials.py

```diff
@@ -1143,3 +1143,28 @@ def add_credential_response(self, stubber):
             }
         }
         stubber.add_response(body=json.dumps(response).encode('utf-8'))
+
+
+class TestContextCredentials(unittest.TestCase):
+    ACCESS_KEY = "access-key"
+    SECRET_KEY = "secret-key"
+
+    def _add_fake_creds(self, request, **kwargs):
+        request.context.setdefault('signing', {})
+        request.context['signing']['request_credentials'] = Credentials(
+            self.ACCESS_KEY, self.SECRET_KEY
+        )
+
+    def _verify_bucket_and_key_in_context(self, request, **kwargs):
+        assert self.ACCESS_KEY in request.headers.get('Authorization')
+
+    def test_credential_context_override(self):
+        session = StubbedSession()
+        with SessionHTTPStubber(session) as stubber:
+            s3 = session.create_client('s3')
+            s3.meta.events.register('before-sign', self._add_fake_creds)
+            s3.meta.events.register(
+                'request-created', self._verify_bucket_and_key_in_context
+            )
```

Comment:
Instead of adding our test as an event, can we make an assertion on the end request instead? That should be possible to check with the stubber while avoiding ambiguity and regressions in the event system.